File: server/lib/routes/balances.dart
import 'dart:convert';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import '../stripe_client.dart';
import '../store/json_store.dart';

Router balanceRoutes() {
  final router = Router();

  // Return locally cached balance data
  router.get('/api/balances', (Request request) {
    final balances = loadBalances();
    final accounts = readAccountData();

    final result = <Map<String, dynamic>>[];
    for (final acct in accounts.accounts) {
      final cached = balances[acct.id] as Map<String, dynamic>?;
      result.add({
        'account_id': acct.id,
        'institution': acct.institution,
        'display_name': acct.displayName,
        'last4': acct.last4,
        'balance': cached?['balance'],
        'balance_refresh': cached?['balance_refresh'],
        'last_refreshed': cached?['last_refreshed'],
        'error': cached?['error'],
      });
    }

    return Response.ok(
      jsonEncode(result),
      headers: {'Content-Type': 'application/json'},
    );
  });

  // Trigger balance refresh via Stripe API — costs ~$0.10 per account
  router.post('/api/balances/refresh', (Request request) async {
    try {
      final body = await request.readAsString();
      Map<String, dynamic>? json;
      if (body.isNotEmpty) {
        json = jsonDecode(body) as Map<String, dynamic>?;
      }

      final accounts = readAccountData();
      final requestedIds = json?['accountIds'] as List<dynamic>?;

      final toRefresh = requestedIds != null
          ? accounts.accounts
              .where((a) => requestedIds.contains(a.id))
              .toList()
          : accounts.accounts;

      if (toRefresh.isEmpty) {
        return Response.ok(
          jsonEncode({'error': 'No accounts to refresh'}),
          headers: {'Content-Type': 'application/json'},
        );
      }

      final balances = loadBalances();
      final results = <Map<String, dynamic>>[];

      for (final acct in toRefresh) {
        try {
          // Trigger balance refresh
          await stripeClient.refreshAccount(acct.id, ['balance']);

          // Poll until refresh completes (max 30 attempts, ~60s)
          Map<String, dynamic>? accountData;
          for (var i = 0; i < 30; i++) {
            await Future.delayed(const Duration(seconds: 2));
            accountData = await stripeClient.getAccount(acct.id);

            final refreshStatus = accountData['balance_refresh']
                as Map<String, dynamic>?;
            if (refreshStatus == null) break;

            final status = refreshStatus['status'] as String?;
            if (status == 'succeeded' || status == 'failed') break;
          }

          if (accountData != null) {
            final entry = {
              'balance': accountData['balance'],
              'balance_refresh': accountData['balance_refresh'],
              'last_refreshed': DateTime.now().toUtc().toIso8601String(),
            };
            balances[acct.id] = entry;

            results.add({
              'account_id': acct.id,
              'institution': acct.institution,
              'display_name': acct.displayName,
              'last4': acct.last4,
              ...entry,
            });
          }
        } on StripeException catch (e) {
          final entry = {
            'error': e.message,
            'last_refreshed': DateTime.now().toUtc().toIso8601String(),
          };
          balances[acct.id] = entry;

          results.add({
            'account_id': acct.id,
            'institution': acct.institution,
            'display_name': acct.displayName,
            'last4': acct.last4,
            ...entry,
          });
        }
      }

      saveBalances(balances);

      return Response.ok(
        jsonEncode(results),
        headers: {'Content-Type': 'application/json'},
      );
    } catch (e) {
      return Response.internalServerError(
        body: jsonEncode({'error': e.toString()}),
        headers: {'Content-Type': 'application/json'},
      );
    }
  });

  return router;
}

File: server/src/routes/balances.ts
import { Hono } from 'hono';
import { stripe } from '../stripe.js';
import {
  loadBalances,
  readAccountData,
  saveBalances,
} from '../store/jsonStore.js';

const app = new Hono();

app.get('/api/balances', (c) => {
  const balances = loadBalances();
  const { accounts } = readAccountData();

  const result = accounts.map((acct) => {
    const cached = balances[acct.id] as Record<string, unknown> | undefined;
    return {
      account_id: acct.id,
      institution: acct.institution,
      display_name: acct.display_name,
      last4: acct.last4,
      balance: cached?.['balance'],
      balance_refresh: cached?.['balance_refresh'],
      last_refreshed: cached?.['last_refreshed'],
      error: cached?.['error'],
    };
  });

  return c.json(result);
});

app.post('/api/balances/refresh', async (c) => {
  try {
    let requestedIds: string[] | undefined;
    try {
      const body = await c.req.json<{ accountIds?: string[] }>();
      requestedIds = body.accountIds;
    } catch {
      // No body
    }

    const accountData = readAccountData();
    const toRefresh =
      requestedIds && requestedIds.length > 0
        ? accountData.accounts.filter((a) => requestedIds!.includes(a.id))
        : accountData.accounts;

    if (toRefresh.length === 0) {
      return c.json({ error: 'No accounts to refresh' });
    }

    const balances = loadBalances() as Record<string, Record<string, unknown>>;
    const results: Record<string, unknown>[] = [];

    for (const acct of toRefresh) {
      try {
        await stripe.financialConnections.accounts.refresh(acct.id, {
          features: ['balance'],
        });

        let accountData: Record<string, unknown> | undefined;
        for (let i = 0; i < 30; i++) {
          await new Promise((resolve) => setTimeout(resolve, 2000));
          const retrieved = await stripe.financialConnections.accounts.retrieve(acct.id);
          accountData = retrieved as unknown as Record<string, unknown>;

          const refreshStatus = accountData['balance_refresh'] as { status?: string } | null;
          if (!refreshStatus) break;
          const status = refreshStatus.status;
          if (status === 'succeeded' || status === 'failed') break;
        }

        if (accountData) {
          const entry: Record<string, unknown> = {
            balance: accountData['balance'],
            balance_refresh: accountData['balance_refresh'],
            last_refreshed: new Date().toISOString(),
          };
          balances[acct.id] = entry;
          results.push({
            account_id: acct.id,
            institution: acct.institution,
            display_name: acct.display_name,
            last4: acct.last4,
            ...entry,
          });
        }
      } catch (err: unknown) {
        const stripeErr = err as { message?: string };
        const entry: Record<string, unknown> = {
          error: stripeErr?.message ?? String(err),
          last_refreshed: new Date().toISOString(),
        };
        balances[acct.id] = entry;
        results.push({
          account_id: acct.id,
          institution: acct.institution,
          display_name: acct.display_name,
          last4: acct.last4,
          ...entry,
        });
      }
    }

    saveBalances(balances);
    return c.json(results);
  } catch (err) {
    return c.json({ error: String(err) }, 500);
  }
});

export default app;
